### ISSUE TITLE
Explore QAT support in TRL

### ISSUE THREAD

**Issue by qgallouedec** [huggingface/trl#6567]
TRL only does *post-training* quantization today: `get_quantization_config` returns a `BitsAndBytesConfig`, so the model is frozen-quantized then adapted with LoRA (QLoRA). No quantization-aware training anywhere.

Question: is QAT worth supporting; fake-quant during the forward pass so the model learns to absorb quantization error, then convert to a real quantized checkpoint?

Why now?: the big releases ship QAT, not PTQ. [Kimi K3](https://www.kimi.com/blog/kimi-k3) applies it from SFT onward (MXFP4 weights, MXFP8 activations) and K2-Thinking shipped INT4 W4A16. [Gemma 4](https://blog.google/innovation-and-ai/technology/developers-tools/quantization-aware-training-gemma-4/) released QAT checkpoints for every variant, including *unquantized* ones meant to be fine-tuned then re-quantize. And [gpt-oss](https://huggingface.co/docs/transformers/quantization/mxfp4) is trained natively in MXFP4, yet `examples/scripts/sft_gpt_oss.py` must pass `dequantize=True` to fine-tune it: we hand back a bf16 model that needs PTQ to redeploy.

To look into:
- Backend: `torchao.quantization.qat` — `quantize_(model, QATConfig(base_config, step="prepare"))`, train, then the same call with `step="convert"`. Axolotl and Unsloth both integrate it; torchao is not yet a TRL dependency.
- Where it lives: `TorchAoConfig` is load-time PTQ only, but `HfQuantizer` already exposes `is_qat_trainable`, honoured by `compressed_tensors`, `bitnet` and `fp_quant`; not torchao. Upstreaming the prepare/convert plumbing there and leaving TRL a config flag looks better than building it locally. Worth asking the transformers quantization folks first.
- Convert is a post-training step we don't have today (Axolotl ships a separate `quantize` command), and `fake_quant_after_n_steps` (delay fake-quant for N steps) needs a training-loop hook -> a callback?
- Which trainers: SFT first, where K3 starts and no rollout story is needed. For GRPO, fake-quant hits only the training forward while vLLM generates from unquantized weights; [slime/SGLang](https://www.lmsys.org/blog/2026-01-26-int4-qat/) quantize the rollout too and report train–infer logprob gaps overlapping BF16. A design question, not a blocker, and it lands on the train–infer mismatch we already care about.
- PEFT: QAT + LoRA vs full-parameter QAT.
- Export: does a converted model round-trip through `save_pretrained` / Hub / vLLM? torchao safetensors serialization exists (>= 0.15) but is documented for the `TorchAoConfig` path, not `quantize_`.

Outcome wanted: a yes/no on feasibility plus a sketch of the smallest useful version, not an implementation.

Related: https://github.com/huggingface/trl/issues/3399, https://github.com/huggingface/transformers/issues/10639

**Comment by DaoyuanLi2816** [comment]
I tested the smallest path with torchao 0.17, Transformers 5.14.1, and a tiny Qwen2 model.

**Verdict:** QAT is feasible for SFT, but I would keep conversion out of the trainer for the first version. The useful initial contract is: train with fake quantization, save a normal high-precision checkpoint, then quantize that checkpoint in a separate export step.

What worked in the probe:

- Applying `quantize_(model, QATConfig(Int4WeightOnlyConfig(...), step="prepare"))` replaced the eligible linears with `FakeQuantizedLinear`, and a full-parameter AdamW step completed with a finite loss.
- `save_pretrained` on the prepared model produced an ordinary safetensors checkpoint with no fake-quantizer state keys. A normal Transformers reload worked; reapplying `prepare` before loading the state also restored a trainable QAT model.
- A PEFT LoRA wrapper on the prepared model also completed a step with only LoRA parameters trainable. `merge_and_unload`, save, and normal reload worked in the toy case. I would still treat this as follow-up scope because the intended LoRA/QAT merge and export semantics need explicit tests.

The main gap is conversion/export rather than training. Direct torchao conversion on this CPU environment was blocked by the optional MSLK backend required by the int4 layout. More importantly, the existing Transformers `TorchAoHfQuantizer` is a load-time PTQ path and does not opt into `is_qat_trainable`; its serialization support should not be assumed to cover an in-place QAT-prepared model.

My proposed smallest version:

1. Support SFT with full-parameter QAT only, initially for torchao's supported `Int4WeightOnlyConfig`.
2. Load a normal high-precision model, apply QAT preparation before Accelerator/FSDP wraps it, and let the existing trainer/checkpoint loop run unchanged.
3. On resume, rebuild the prepared module structure before loading the checkpoint.
4. Save the trained high-precision weights normally. Provide a separate export example/command that reloads them with `TorchAoConfig(base_config)` and then uses the existing Transformers save path.
5. Add one CUDA integration test covering prepare -> train step -> save -> reload with `TorchAoConfig` -> quantized forward. The CPU test can cover prepare, backward, checkpoint, and resume.

I would leave delayed fake quantization, PEFT, and GRPO out of v1. Delayed activation can later be a callback using torchao's fake-quant enable/disable helpers. GRPO needs a separate train/inference parity design because preparing only the training model would not address the vLLM side.

This does not require an `HfQuantizer` change for the initial training path: the model starts unquantized and is prepared after loading. Longer term, moving the lifecycle into Transformers would make sense if torchao exposes a stable QAT-aware config and conversion/serialization contract.

References: [torchao QAT workflow](https://docs.pytorch.org/ao/stable/workflows/qat.html), [torchao finetuning tutorial](https://docs.pytorch.org/ao/stable/eager_tutorials/finetuning.html), [Transformers torchao integration](https://huggingface.co/docs/transformers/quantization/torchao).

**Comment by opertifelipe** [comment]
I did some additional experiments around the smallest useful SFT integration and put together a prototype implementation in my fork:

[https://github.com/opertifelipe/trl/pull/1](https://github.com/opertifelipe/trl/pull/1)

The prototype currently focuses on full-parameter TorchAO INT4 weight-only QAT in `SFTTrainer`. It prepares the model before Trainer/Accelerate wrapping, keeps the existing training loop unchanged, supports checkpoint/resume by recreating the QAT module topology, and leaves the actual INT4 conversion as an explicit post-training step.

I also added tests covering QAT preparation, forward/backward with finite gradients, an actual optimizer update, invalid configurations, checkpoint resume, and a CUDA smoke test for conversion followed by quantized inference.

The training side has been relatively straightforward. The conversion/export side is more backend- and hardware-dependent, because the usable INT4 inference configuration depends on the target hardware, the PyTorch/TorchAO builds, and the available kernels.

On systems where TorchAO's default `plain` packing format and `tinygemm` qparams algorithm are supported, the deployment conversion can use the minimal configuration:

```python
Int4WeightOnlyConfig(group_size=128)
```

On some CUDA setups, however, I found that conversion and inference required an explicit deployment configuration such as:

```python
Int4WeightOnlyConfig(
    group_size=128,
    int4_packing_format="tile_packed_to_4d",
    int4_choose_qparams_algorithm="hqq",
)
```

Both configurations preserve the `group_size=128` granularity used during QAT while selecting different deployment paths. With the latter configuration, I was able to convert the QAT-trained model to packed INT4 weights and successfully run an inference pass.

Because of this hardware/backend sensitivity, I still consider conversion/export something that should be validated on the target deployment environment rather than hard-coded into the Trainer. In particular, I would not assume that different TorchAO packing/qparam configurations are bit-for-bit equivalent in numerics or serialization behavior, even when they preserve the same quantization granularity.

I kept PEFT/LoRA, QLoRA, delayed fake quantization, and rollout-based trainers out of scope for this first prototype.

This is currently just a draft implementation in my fork to explore the design rather than an upstream PR. If this direction looks useful for TRL, I'd be happy to continue working on it, investigate the remaining TorchAO conversion/export behavior across supported backends, adapt the design based on maintainer feedback, and develop the upstream contribution.
